generator/sponsored_dataset_inl.hpp

```diff
@@ -0,0 +1,179 @@
+
+  for (platform::LocalCountryFile const & localFile : localFiles)
+  {
+    LOG(LINFO, ("Found mwm:", localFile));
+    try
+    {
+      m_index.RegisterMap(localFile);
+    }
+    catch (RootException const & ex)
+    {
+      CHECK(false, (ex.Msg(), "Bad mwm file:", localFile));
+    }
+  }
+
+  m_coder = make_unique<search::ReverseGeocoder>(m_index);
+}
+
+template <typename SponsoredObject>
+void SponsoredDataset<SponsoredObject>::AddressMatcher::operator()(Object & object)
+{
+  search::ReverseGeocoder::Address addr;
+  m_coder->GetNearbyAddress(MercatorBounds::FromLatLon(object.m_latLon), addr);
+  object.m_street = addr.GetStreetName();
+  object.m_houseNumber = addr.GetHouseNumber();
+}
+
+
+// SponsoredDataset --------------------------------------------------------------------------------
+template <typename SponsoredObject>
+SponsoredDataset<SponsoredObject>::SponsoredDataset(string const & dataPath, string const & addressReferencePath)
+{
+  if (dataPath.empty())
+    return;
+
+  ifstream dataSource(dataPath);
+  if (!dataSource.is_open())
+  {
+    LOG(LERROR, ("Error while opening", dataPath, ":", strerror(errno)));
+    return;
+  }
+
+  LoadData(dataSource, addressReferencePath);
+}
+
+template <typename SponsoredObject>
+SponsoredDataset<SponsoredObject>::SponsoredDataset(istream & dataSource, string const & addressReferencePath)
+{
+  LoadData(dataSource, addressReferencePath);
+}
+
+template <typename SponsoredObject>
+typename SponsoredDataset<SponsoredObject>::Object const &
+SponsoredDataset<SponsoredObject>::GetObjectById(ObjectId id) const
+{
+  auto const it = m_objects.find(id);
+  CHECK(it != end(m_objects), ("Got wrong object id:", id));
+  return it->second;
+}
+
+template <typename SponsoredObject>
+typename SponsoredDataset<SponsoredObject>::Object &
+SponsoredDataset<SponsoredObject>::GetObjectById(ObjectId id)
+{
+  auto const it = m_objects.find(id);
+  CHECK(it != end(m_objects), ("Got wrong object id:", id));
+  return it->second;
+}
+
+template <typename SponsoredObject>
+void SponsoredDataset<SponsoredObject>::BuildOsmObjects(function<void(FeatureBuilder1 &)> const & fn) const
+{
+  for (auto const & item : m_objects)
+    BuildObject(item.second, fn);
+}
+
+template <typename SponsoredObject>
+typename SponsoredDataset<SponsoredObject>::ObjectId
+SponsoredDataset<SponsoredObject>::FindMatchingObjectId(FeatureBuilder1 const & fb) const
+{
+  if (NecessaryMatchingConditionHolds(fb))
+    return FindMatchingObjectIdImpl(fb);
+  return Object::InvalidObjectId();
+}
+
+template <typename SponsoredObject>
+vector<typename SponsoredDataset<SponsoredObject>::ObjectId>
+SponsoredDataset<SponsoredObject>::GetNearestObjects(ms::LatLon const & latLon, size_t const limit,
+                                                     double const maxDistanceMeters /* = 0.0 */) const
+{
+  namespace bgi = boost::geometry::index;
+
+  vector<ObjectId> indexes;
```

Comment:
We use r-tree here with 'nearest' predicate, as KNN-index. I think geometry/tree4d not a good idea.